### PR TITLE
[jvm-packages] resolve spark compatibility issue

### PR DESCRIPTION
**Pull request by wbo4958**
The latest xgboost couldn't run on older spark versions due to some APIs newly added in latest Spark.

This PR try to fix this issue.

I ran the xgboost4j-spark test on `spark 3.1.3/3.2.4/3.3.4/3.4.3/` without any issue.

### CODE REVIEW

**Comment by wbo4958** [comment]
Hi @hcho3,

Could you help add below tests when compiling xgboost4j-spark just for CPU?


``` bash
mvn clean package -Dspark.version=3.1.3 -pl xgboost4j-spark
mvn clean package -Dspark.version=3.2.4 -pl xgboost4j-spark
mvn clean package -Dspark.version=3.3.4 -pl xgboost4j-spark
mvn clean package -Dspark.version=3.4.3 -pl xgboost4j-spark
```

**Comment by wbo4958** [comment]
Hi @hcho3, there is an issue that flink tests will hang forever when we specify the spark version to be like 3.2.1,

For example,

`mvn clean package -Dspark.version=3.2.1`, I've no idea why spark version will affect flink tests. that's really strange. Given that I don't know flink, so the test should pass flink tests by specifying xgboost4j-spark project explicitly. Thx

``` bash
mvn clean package -Dspark.version=3.1.3 -pl xgboost4j-spark
mvn clean package -Dspark.version=3.2.4 -pl xgboost4j-spark
mvn clean package -Dspark.version=3.3.4 -pl xgboost4j-spark
mvn clean package -Dspark.version=3.4.3 -pl xgboost4j-spark
```

**Review by hcho3 — APPROVED** [review]

File: jvm-packages/xgboost4j-spark/src/main/scala/org/apache/spark/ml/xgboost/SparkUtils.scala [hunk 1]
```diff
@@ -57,8 +58,52 @@ trait XGBProbabilisticClassifierParams[T <: Params]
...
+          .otherwise(casted)
+
+      case _ =>
+        val n = numClasses.getOrElse(Int.MaxValue)
+        require(0 < n && n <= Int.MaxValue)
+        when(casted.isNull || casted.isNaN, raise_error(lit("Labels MUST NOT be Null or NaN")))
+          .when(casted < 0 || casted >= n,
+            raise_error(concat(lit(s"Labels MUST be in [0, $n), but got "), casted)))
+          .when(casted =!= casted.cast(IntegerType),
+            raise_error(concat(lit("Labels MUST be Integers, but got "), casted)))
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply jvm-packages/xgboost4j-spark/src/main/scala/org/apache/spark/ml/xgboost/SparkUtils.scala:79]: Yes. the label must be an integer. And this piece of code is copied from spark.

**Comment by wbo4958** [comment]
Hi @trivialfis, could you help review it?

**Review by trivialfis — APPROVED** [review]

**Pull request merged**